packages/protocol/contracts/L2/1559/Lib1559Math.sol

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.20;
+
+import { LibFixedPointMath } from "../../thirdparty/LibFixedPointMath.sol";
+
+library Lib1559Math {
+    error EIP1559_INPUT_TOO_BIG();
+    error EIP1559_INVALID_PARAMS();
+
+    function calcBaseFee(
+        uint64 numL1Blocks,
+        uint128 gasExcess,
+        uint32 gasInBlock,
+        uint64 gasTarget,
+        uint64 adjustmentQuotient
+    )
+        internal
+        pure
+        returns (uint256 _baseFeePerGas, uint128 _gasExcess)
+    {
+        if (gasTarget == 0 || adjustmentQuotient <= 1 || gasInBlock == 0) {
+            revert EIP1559_INVALID_PARAMS();
+        }
+
+        uint128 issuance = numL1Blocks * gasTarget;
+        _gasExcess = issuance >= gasExcess ? 0 : gasExcess - issuance;
+
+        _baseFeePerGas =
+            _calcBaseFee(_gasExcess, gasInBlock, gasTarget, adjustmentQuotient);
+        _gasExcess += gasInBlock;
+    }
+
+    function _calcBaseFee(
+        uint256 gasExcess,
+        uint256 gasInBlock,
+        uint256 gasTarget,
+        uint256 adjustmentQuotient
+    )
+        private
+        pure
+        returns (uint256 baseFeePerGas)
+    {
+        int256 diff = _ethQty(
+            gasExcess + gasInBlock, gasTarget, adjustmentQuotient
+        ) - _ethQty(gasExcess, gasTarget, adjustmentQuotient);
+        baseFeePerGas =
+            uint256(diff) / gasInBlock / LibFixedPointMath.SCALING_FACTOR_1E18;
```

Comment:
I think doing the derivative at the current `gasExcess` is a bit more accurate like how Vitalik does it, because we're working on the state after the previous block (and so we are not calculating the slope for the current block, in which case calculating the diff like is done here would be more accureate). But I guess the biggest thing is that it supports 0 gas blocks and the accuracy is not dependent on how big `gasInBlock` is (UNTESTED so I may do the scaling wrong):
```
baseFeePerGas = _ethQty(gasExcess, gasTarget, adjustmentQuotient) / (gasTarget * adjustmentQuotient * LibFixedPointMath.SCALING_FACTOR_1E18)
```

Reply:
Will test it out

Reply:
It turns out the derivative based base fee calc is more accurate:

```solidity
def basefee1(gas_used):
    diff = eth_qty(gas_excess_issued + gas_used) - eth_qty(gas_excess_issued)
    return int(diff / gas_used /SCALE)

def basefee2(): //  turns out to be better
    return eth_qty(gas_excess_issued) / SCALE / TAIKO_TARGET / ADJUSTMENT_QUOTIENT

```